src/wallet/wallet2.cpp

```diff
@@ -7761,6 +7761,194 @@ bool wallet2::check_tx_proof(const crypto::hash &txid, const cryptonote::account
   return false;
 }
 
+std::string wallet2::get_reserve_proof(const boost::optional<std::pair<uint32_t, uint64_t>> &account_minreserve, const std::string &message)
+{
+  THROW_WALLET_EXCEPTION_IF(m_watch_only || m_multisig, error::wallet_internal_error, "Reserve proof can only be generated by a full wallet");
+  THROW_WALLET_EXCEPTION_IF(account_minreserve && balance(account_minreserve->first) < account_minreserve->second, error::wallet_internal_error,
+    "Not enough balance in this account for the requested minimum reserve amount");
+
+  std::vector<size_t> selected_transfers;
+  for (size_t i = 0; i < m_transfers.size(); ++i)
+  {
+    const transfer_details &td = m_transfers[i];
+    if (!td.m_spent && (!account_minreserve || account_minreserve->first == td.m_subaddr_index.major))
+      selected_transfers.push_back(i);
+  }
+
+  if (account_minreserve)
+  {
+    // minimize the number of outputs included in the proof, by only picking the N largest outputs that can cover the requested min reserve amount
+    std::sort(selected_transfers.begin(), selected_transfers.end(), [&](const size_t a, const size_t b)
+      { return m_transfers[a].amount() > m_transfers[b].amount(); });
+    size_t sz = 0;
+    uint64_t total = 0;
+    while (total < account_minreserve->second)
+    {
+      total += m_transfers[selected_transfers[sz]].amount();
+      ++sz;
+    }
+    selected_transfers.resize(sz);
```

Comment:
While this should not happen with the current code, I'd like a check sz doesn't go beyond selected_transfers.size() here. Technically, it *could* break if a refresh happens and an out gets spent just before the loop (hence the need for LOCK_IDLE_SCOPE above).

Reply:
I think there's some misunderstanding. The spent status is checked above L7931, but ultimately it doesn't matter whether or not any out gets spent right before this loop starts. The set `selected_transfers` remains unaffected, and it's okay that it contains any outs that got spent just now; the resulting signature will be valid, with those outputs being simply displayed as spent.

Reply:
m_transfers can *shrink* in a reorg. It's a vector, so it can also be reallocated when grown so the m_transfers access could break even with a valid index.

Reply:
Do you mean I should check `selected_transfers[i] < m_transfers.size()` every time I do `m_transfers[selected_transfers[i]]`? I'm confused.

Reply:
I mainly mean you should not allow another thread to write to m_transfers while you're doing so.
My comment about testing the index is that the place that fills selected_transfers and the place that uses it are disconnected enough that I didn't feel super confident it'd always stay right, but if you think it's fine without a test, it's OK. I definitely think you should call LOCK_IDLE_SCOPE though.

Reply:
I did incorporate LOCK_IDLE_SCOPE as suggested. (BTW I don't see the same mechanism for the wallet RPC or API. Is this for CLI only?)

I see this idiom quite frequently in wallet2.cpp:
```c++
    for (size_t idx: selected_transfers)
    {
      const transfer_details &td = m_transfers[idx];
      ...
    }
```
Do you imply that they might also potentially crash because `m_transfers` may shrink concurrently and `idx` becomes out of bound?

Reply:
simplewallet has a refresh thread, so it does that. Anything which calls wallet2 in two threads should also do it. This idiom may crash if two threads are running wallet2 concurrently and one of the threads writes to m_transfers.

Reply:
Then, it should be safe as long as LOCK_IDLE_SCOPE is called on simplewallet's side, right? If so, I'd leave the code as is.